backend-services/user-verification-service/src/main/java/org/cardano/foundation/voting/service/discord/DefaultDiscordUserVerificationService.java

```diff
@@ -248,7 +295,124 @@ public Either<Problem, IsVerifiedResponse> checkVerification(DiscordCheckVerific
 
         var pendingVerification = maybePendingVerification.get();
         boolean isSecretCodeMatch = pendingVerification.getSecretCode().equals(cip30Secret)
-                && pendingVerification.getSecretCode().equals(requestSecret);
+                && pendingVerification.getSecretCode().equals(request.getSecret());
+
+        if (!isSecretCodeMatch) {
+            return Either.left(Problem.builder()
+                    .withTitle("AUTH_FAILED")
+                    .withDetail("Invalid secret and / or discordIdHash.")
+                    .withStatus(BAD_REQUEST)
+                    .build()
+            );
+        }
+
+        var pendingUserVerification = maybePendingVerification.orElseThrow();
+
+        var now = LocalDateTime.now(clock);
+
+        var isCodeExpired = now.isAfter(pendingUserVerification.getExpiresAt());
+        if (isCodeExpired) {
+            return Either.left(Problem.builder()
+                    .withTitle("VERIFICATION_EXPIRED")
+                    .withDetail(String.format("Secret code: %s expired for walletId: %s and discordHashId:%s", cip30Secret, walletId, discordIdHash))
+                    .withStatus(BAD_REQUEST)
+                    .with("discordIdHash", discordIdHash)
+                    .with("walletId", walletId)
+                    .build());
+        }
+
+        pendingVerification.setWalletId(Optional.of(request.getWalletId()));
+        pendingVerification.setWalletIdType(request.getWalletIdType());
+        pendingVerification.setUpdatedAt(LocalDateTime.now(clock));
+        pendingVerification.setStatus(VERIFIED);
+        userVerificationRepository.save(pendingVerification);
+
+        return Either.right(new IsVerifiedResponse(true));
+    }
+
+    private Either<Problem, Tuple2<String, Optional<String>>> verifySignature(String signature, String publicKey) {
+        CIP30Verifier verifier = new CIP30Verifier(signature, publicKey);
+        var result = verifier.verify();
+
+        if (!result.isValid()) {
+            return Either.left(Problem.builder()
+                    .withTitle("INVALID_CIP-30-SIGNATURE")
+                    .withDetail("Invalid CIP-30 signature")
+                    .withStatus(BAD_REQUEST)
+                    .build()
+            );
+        }
+
+        String msg = result.getMessage(MessageFormat.TEXT);
+        Optional<String> maybeAddress = result.getAddress(AddressFormat.TEXT);
+
+        return Either.right(new Tuple2<>(msg, maybeAddress));
+    }
+
+    private Either<Problem, IsVerifiedResponse> handleKeriVerification(DiscordCheckVerificationRequest request, String eventId, String walletId) {
```

Comment:
I got confused here a bit, you could help me in the future with adding a spring annotation like this for these cases:
```
    @Transactional(propagation = SUPPORTS)
```

Reply:
done:)